File: ext/sizes/sizes.c
#include "ruby.h"

#if HAVE_STDDEF_H
  #include <stddef.h>
#endif

#if HAVE_STDLIB_H
  #include <stdlib.h>
#endif

#if HAVE_INTTYPES_H
  #include <inttypes.h>
#endif

/* I almost wonder if I should have just generated this file with Ruby. Ah, well... */

static VALUE sizeof_char(VALUE self)               { return INT2FIX(sizeof(char)); }
static VALUE sizeof_unsigned_char(VALUE self)      { return INT2FIX(sizeof(unsigned char)); }
static VALUE sizeof_signed_char(VALUE self)        { return INT2FIX(sizeof(signed char)); }
static VALUE sizeof_signed_int(VALUE self)         { return INT2FIX(sizeof(signed int)); }
static VALUE sizeof_signed_short_int(VALUE self)   { return INT2FIX(sizeof(signed short int)); }
static VALUE sizeof_long_int(VALUE self)           { return INT2FIX(sizeof(long int)); }
static VALUE sizeof_unsigned_int(VALUE self)       { return INT2FIX(sizeof(unsigned int)); }
static VALUE sizeof_unsigned_short_int(VALUE self) { return INT2FIX(sizeof(unsigned short int)); }
static VALUE sizeof_unsigned_long_int(VALUE self)  { return INT2FIX(sizeof(unsigned long int)); }
static VALUE sizeof_float(VALUE self)              { return INT2FIX(sizeof(float)); }
static VALUE sizeof_double(VALUE self)             { return INT2FIX(sizeof(double)); }
static VALUE sizeof_long_double(VALUE self)        { return INT2FIX(sizeof(long double)); }

/* C99 types */
#if HAVE_STDDEF_H || HAVE_STDLIB_H
  static VALUE sizeof_size(VALUE self)        { return INT2FIX(sizeof(size_t)); }
  static VALUE sizeof_wchar(VALUE self)       { return INT2FIX(sizeof(wchar_t)); }
#endif

#if HAVE_INTTYPES_H
  static VALUE sizeof_int8(VALUE self)          { return INT2FIX(sizeof(int8_t));   }
  static VALUE sizeof_int16(VALUE self)         { return INT2FIX(sizeof(int16_t));  }
  static VALUE sizeof_int32(VALUE self)         { return INT2FIX(sizeof(int32_t));  }
  static VALUE sizeof_int64(VALUE self)         { return INT2FIX(sizeof(int64_t));  }
  static VALUE sizeof_uint8(VALUE self)         { return INT2FIX(sizeof(uint8_t));  }
  static VALUE sizeof_uint16(VALUE self)        { return INT2FIX(sizeof(uint16_t)); }
  static VALUE sizeof_uint32(VALUE self)        { return INT2FIX(sizeof(uint32_t)); }
  static VALUE sizeof_uint64(VALUE self)        { return INT2FIX(sizeof(uint64_t)); }

  static VALUE sizeof_int_least8(VALUE self)    { return INT2FIX(sizeof(int_least8_t));   }
  static VALUE sizeof_int_least16(VALUE self)   { return INT2FIX(sizeof(int_least16_t));  }
  static VALUE sizeof_int_least32(VALUE self)   { return INT2FIX(sizeof(int_least32_t));  }
  static VALUE sizeof_int_least64(VALUE self)   { return INT2FIX(sizeof(int_least64_t));  }
  static VALUE sizeof_uint_least8(VALUE self)   { return INT2FIX(sizeof(uint_least8_t));  }
  static VALUE sizeof_uint_least16(VALUE self)  { return INT2FIX(sizeof(uint_least16_t)); }
  static VALUE sizeof_uint_least32(VALUE self)  { return INT2FIX(sizeof(uint_least32_t)); }
  static VALUE sizeof_uint_least64(VALUE self)  { return INT2FIX(sizeof(uint_least64_t)); }

  static VALUE sizeof_int_fast8(VALUE self)     { return INT2FIX(sizeof(int_fast8_t));   }
  static VALUE sizeof_int_fast16(VALUE self)    { return INT2FIX(sizeof(int_fast16_t));  }
  static VALUE sizeof_int_fast32(VALUE self)    { return INT2FIX(sizeof(int_fast32_t));  }
  static VALUE sizeof_int_fast64(VALUE self)    { return INT2FIX(sizeof(int_fast64_t));  }
  static VALUE sizeof_uint_fast8(VALUE self)    { return INT2FIX(sizeof(uint_fast8_t));  }
  static VALUE sizeof_uint_fast16(VALUE self)   { return INT2FIX(sizeof(uint_fast16_t)); }
  static VALUE sizeof_uint_fast32(VALUE self)   { return INT2FIX(sizeof(uint_fast32_t)); }
  static VALUE sizeof_uint_fast64(VALUE self)   { return INT2FIX(sizeof(uint_fast64_t)); }

  static VALUE sizeof_intptr(VALUE self)        { return INT2FIX(sizeof(intptr_t));  }
  static VALUE sizeof_uintptr(VALUE self)       { return INT2FIX(sizeof(uintptr_t)); }
  static VALUE sizeof_intmax(VALUE self)        { return INT2FIX(sizeof(intmax_t));  }
  static VALUE sizeof_uintmax(VALUE self)       { return INT2FIX(sizeof(uintmax_t)); }
#endif

void Init_sizes()
{
  VALUE mSizes = rb_define_module("Sizes");
  
  rb_define_module_function(mSizes, "sizeof_char",               sizeof_char,               0);
  rb_define_module_function(mSizes, "sizeof_unsigned_char",      sizeof_unsigned_char,      0);
  rb_define_module_function(mSizes, "sizeof_signed_char",        sizeof_signed_char,        0);
  rb_define_module_function(mSizes, "sizeof_signed_int",         sizeof_signed_int,         0);
  rb_define_module_function(mSizes, "sizeof_signed_short_int",   sizeof_signed_short_int,   0);
  rb_define_module_function(mSizes, "sizeof_long_int",           sizeof_long_int,           0);
  rb_define_module_function(mSizes, "sizeof_unsigned_int",       sizeof_unsigned_int,       0);
  rb_define_module_function(mSizes, "sizeof_unsigned_short_int", sizeof_unsigned_short_int, 0);
  rb_define_module_function(mSizes, "sizeof_unsigned_long_int",  sizeof_unsigned_long_int,  0);
  rb_define_module_function(mSizes, "sizeof_float",              sizeof_float,              0);
  rb_define_module_function(mSizes, "sizeof_double",             sizeof_double,             0);
  rb_define_module_function(mSizes, "sizeof_long_double",        sizeof_long_double,        0);
  
  /* C99 types */
  #if HAVE_STDDEF_H || HAVE_STDLIB_H
    rb_define_module_function(mSizes, "sizeof_size",        sizeof_size,        0);
    rb_define_module_function(mSizes, "sizeof_wchar",       sizeof_wchar,       0);
  #endif
  
  #if HAVE_STDINT_H
    rb_define_module_function(mSizes, "sizeof_int8",        sizeof_int8,        0);
    rb_define_module_function(mSizes, "sizeof_int16",       sizeof_int16,       0);
    rb_define_module_function(mSizes, "sizeof_int32",       sizeof_int32,       0);
    rb_define_module_function(mSizes, "sizeof_int64",       sizeof_int64,       0);
    rb_define_module_function(mSizes, "sizeof_uint8",       sizeof_uint8,       0);
    rb_define_module_function(mSizes, "sizeof_uint16",      sizeof_uint16,      0);
    rb_define_module_function(mSizes, "sizeof_uint32",      sizeof_uint32,      0);
    rb_define_module_function(mSizes, "sizeof_uint64",      sizeof_uint64,      0);

    rb_define_module_function(mSizes, "sizeof_int_least8",  sizeof_int_least8,  0);
    rb_define_module_function(mSizes, "sizeof_int_least16", sizeof_int_least16, 0);
    rb_define_module_function(mSizes, "sizeof_int_least32", sizeof_int_least32, 0);
    rb_define_module_function(mSizes, "sizeof_int_least64", sizeof_int_least64, 0);
    rb_define_module_function(mSizes, "sizeof_uint_least8", sizeof_uint_least8, 0);
    rb_define_module_function(mSizes, "sizeof_uint_least16",sizeof_uint_least16,0);
    rb_define_module_function(mSizes, "sizeof_uint_least32",sizeof_uint_least32,0);
    rb_define_module_function(mSizes, "sizeof_uint_least64",sizeof_uint_least64,0);

    rb_define_module_function(mSizes, "sizeof_int_fast8",   sizeof_int_fast8,   0);
    rb_define_module_function(mSizes, "sizeof_int_fast16",  sizeof_int_fast16,  0);
    rb_define_module_function(mSizes, "sizeof_int_fast32",  sizeof_int_fast32,  0);
    rb_define_module_function(mSizes, "sizeof_int_fast64",  sizeof_int_fast64,  0);
    rb_define_module_function(mSizes, "sizeof_uint_fast8",  sizeof_uint_fast8,  0);
    rb_define_module_function(mSizes, "sizeof_uint_fast16", sizeof_uint_fast16, 0);
    rb_define_module_function(mSizes, "sizeof_uint_fast32", sizeof_uint_fast32, 0);
    rb_define_module_function(mSizes, "sizeof_uint_fast64", sizeof_uint_fast64, 0);

    rb_define_module_function(mSizes, "sizeof_intptr",      sizeof_intptr,      0);
    rb_define_module_function(mSizes, "sizeof_uintptr",     sizeof_uintptr,     0);
    rb_define_module_function(mSizes, "sizeof_intmax",      sizeof_intmax,      0);
    rb_define_module_function(mSizes, "sizeof_uintmax",     sizeof_uintmax,     0);
  #endif
}

File: lib/sizes.rb
require File.join(File.dirname(__FILE__), "../ext/sizes/sizes")

module Sizes
  module_function
  C99_TYPES = %w(int8 int16 int32 int64 uint8 uint16 uint32 uint64 int_least8 int_least16 int_least32
                 int_least64 uint_least8 uint_least16 uint_least32 uint_least64 int_fast8 int_fast16 int_fast32
                 int_fast64 uint_fast8 uint_fast16 uint_fast32 uint_fast64 intptr uintptr intmax uintmax size
                 wchar)
  
  STRING_COPIES = (
                    %w(char unsigned_char signed_char signed_int int signed_short_int signed_long_int
                       long_int long unsigned_int unsigned unsigned_short_int unsigned_short unsigned_long_int
                       unsigned_long float double long_double) + C99_TYPES
                  ).inject({}) { |hash,key| hash[key] = key.to_sym; hash }.freeze
    
  def sizeof(symbol_or_string)
    # I could just use send("sizeof_#{symbol_or_string}") but that would be slower, and negate any advantage of using
    # symbols. Not sure how much the speed difference really matters, but it couldn't hurt, right?
    
    case symbol_or_string
      when String
        if sym = STRING_COPIES[symbol_or_string]
          sizeof(sym)
        else
          raise ArgumentError, "Argument must be one of #{STRING_COPIES.values.inspect}, or a String version of the same"
        end
      when :char                                 then sizeof_char
      when :unsigned_char                        then sizeof_unsigned_char
      when :signed_char                          then sizeof_signed_char
      when :signed_int, :int                     then sizeof_signed_int
      when :signed_short_int, :short_int, :short then sizeof_signed_short_int
      when :signed_long_int, :long_int, :long    then sizeof_long_int
      when :unsigned_int, :unsigned              then sizeof_unsigned_int
      when :unsigned_short_int, :unsigned_short  then sizeof_unsigned_short_int
      when :unsigned_long_int, :unsigned_long    then sizeof_unsigned_long_int
      when :float                                then sizeof_float
      when :double                               then sizeof_double
      when :long_double                          then sizeof_long_double
      when :int8                                 then sizeof_int8
      when :int16                                then sizeof_int16
      when :int32                                then sizeof_int32
      when :int64                                then sizeof_int64
      when :uint8                                then sizeof_uint8
      when :uint16                               then sizeof_uint16
      when :uint32                               then sizeof_uint32
      when :uint64                               then sizeof_uint64
      when :int_least8                           then sizeof_int_least8
      when :int_least16                          then sizeof_int_least16
      when :int_least32                          then sizeof_int_least32
      when :int_least64                          then sizeof_int_least64
      when :uint_least8                          then sizeof_uint_least8
      when :uint_least16                         then sizeof_uint_least16
      when :uint_least32                         then sizeof_uint_least32
      when :uint_least64                         then sizeof_uint_least64
      when :int_fast8                            then sizeof_int_fast8
      when :int_fast16                           then sizeof_int_fast16
      when :int_fast32                           then sizeof_int_fast32
      when :int_fast64                           then sizeof_int_fast64
      when :uint_fast8                           then sizeof_uint_fast8
      when :uint_fast16                          then sizeof_uint_fast16
      when :uint_fast32                          then sizeof_uint_fast32
      when :uint_fast64                          then sizeof_uint_fast64
      when :intptr                               then sizeof_intptr
      when :uintptr                              then sizeof_uintptr
      when :intmax                               then sizeof_intmax
      when :uintmax                              then sizeof_uintmax
      when :size_t, :size                        then sizeof_size
      when :wchar_t, :wchar                      then sizeof_wchar
      else
        raise ArgumentError, "Argument must be one of #{STRING_COPIES.values.inspect}, or a String version of the same"
    end
  end
end